pyomo/core/base/set.py

```diff
@@ -147,12 +226,26 @@ def A_rule(model, i, j):
         ...
     """
 
-    def wrapper_function ( *args, **kwargs ):
-        value = fn( *args, **kwargs )
+    # Because some of our processing of initializer functions relies on
+    # knowing the number of positional arguments, we will go to extra
+    # effort here to preserve the original function signature.
+    _funcdef = """def wrapper_function%s:
+        args, varargs, kwds, local_env = inspect.getargvalues(
+            inspect.currentframe())
+        args = tuple(local_env[_] for _ in args) + (varargs or ())
+        value = fn(*args, **(kwds or {}))
+        # Map None -> Set.End
         if value is None:
             return Set.End
         return value
-    return wrapper_function
+""" % (formatargspec(fn),)
+    # Create the wrapper in a temporary environment that mimics this
+    # function's environment.
+    _env = dict(globals())
+    _env.update(locals())
+    exec(_funcdef, _env)
+    return _env['wrapper_function']
```

Comment:
I understand the need to preserve the original function signature. However, is a string followed by exec the best way to do this? Maybe I do not quite understand the goal here. Can you expand on this further?

Reply:
I spent a long time on this, and have not been able to find a way to preserve the original function signature without using `exec`.  The challenge is that some of the obscure (and now generally unused) rules for functions added an additional positional argument to the rule that was the number of times the rule had been called (this was legacy support from before we understood the power of generators).  To preserve backwards comparability with that way of calling rules, we need to know how many positional arguments the rule was expecting.  If we create the wrapper in the normal way (with `*args`), that information is lost.  So, when we wrap a rule, we generate a new function with *exactly* the same calling signature.  I consider this reasonably safe, as the string is "just" a function definition (no side effects), and the variable data does not come from the user, but instead from a Python internal function.

Reply:
Okay, I figured you had a good reason, but I had to ask.